core/domain/classifier_services.py

```diff
@@ -100,8 +101,38 @@ def classify_string_classifier_rule(state, normalized_answer):
             training_examples.extend([
                 [doc, [str(answer_group_index)]]
                 for doc in classifier_rule_spec.inputs['training_data']])
-    if len(training_examples) > 0:
+
+    exploration = exp_services.get_exploration_by_id(exp_id)
+    state_obj = exploration.states[state_name]
+    algorithm_id = feconf.INTERACTION_CLASSIFIER_MAPPING['TextInput']
+    if state_obj.classifier_model_id:
+        classifier = get_classifier_by_id(state_obj.classifier_model_id)
+        sc_dict = classifier.cached_classifier_data
+        sc = classifier_registry.Registry.get_classifier_by_algorithm_id(
+            algorithm_id)
+        sc.from_dict(sc_dict)
+
+    else:
+        sc = classifier_registry.Registry.get_classifier_by_algorithm_id(
+            algorithm_id)
         sc.train(training_examples)
+        cached_classifier_data = sc.to_dict()
+        algorithm_version = feconf.CLASSIFIER_VERSION_MAPPING[algorithm_id]
+        classifier = classifier_domain.Classifier(
+            '0', exploration.id, exp_version, state_name, algorithm_id,
```

Comment:
Then it would be better to declare `temporary_classifier_id = '0'` and use that instead, so that readers of the code understand the context.